hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/HazelcastRexBuilder.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.calcite;
+
+import com.hazelcast.sql.impl.calcite.validate.types.HazelcastIntegerType;
+import com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeFactory;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+
+import static org.apache.calcite.sql.type.SqlTypeName.ANY;
+import static org.apache.calcite.sql.type.SqlTypeName.REAL;
+
+/**
+ * Custom Hazelcast expression builder.
+ * <p>
+ * Currently, this custom expression builder is used just to workaround quirks
+ * of the default Calcite expression builder.
+ */
+public final class HazelcastRexBuilder extends RexBuilder {
+
+    public HazelcastRexBuilder(RelDataTypeFactory typeFactory) {
+        super(typeFactory);
+        assert typeFactory instanceof HazelcastTypeFactory;
+    }
+
+    @Override
+    public RexNode makeLiteral(Object value, RelDataType type, boolean allowCast) {
+        // XXX: Calcite evaluates casts like CAST(0 AS ANY) statically and
```

Comment:
What does `CAST X as ANY` mean from the user perspective? We do not have `ANY` type that is exposed to a user, are we?

Reply:
Currently `ANY` is exposed to users. Honestly, I'm not sure how to implement `OBJECT` support properly. `ANY` is the only standard type (`SqlTypeName`) having the semantics similar to `OBJECT`, it's actually mapped to `java.sql.Types.JAVA_OBJECT` by Calcite itself. Yes, it does have a special "skip-all-checks" meaning during validation. Probably SQL user defined types may help (not sure how well they are supported by Calcite), but AFAIU user defined types can express a new type only in terms of well-known ones, so internally we need to map it to some existing type having `OBJECT` semantics. Probably adding a custom `RelDataType` may resolve that, not sure. This requires research. What exactly you don't like about using `ANY`, the fact that we refer to it in queries as "ANY" and not as "OBJECT" or something else?

Reply:
Probably I'm not getting right what the expected semantics of `OBJECT` actually is. Does it represent an object with nested fields accessible from queries or just a raw Java object?

Reply:
This is up to us to decide in the scope of this PR because it was not a concern before. 

In the prototype branch, the original semantics was "this is an object of some type, I do not know which exactly". So perhaps `ANY` is a good candidate. In this case, `SqlColumnType` should be renamed accordingly.

On the other hand, `ANY` maybe not an ideal fit for DDL, because in other databases `ANY` counterparts mean `it could store anything`, while in our case this is definitely not the case - one cannot store arbitrary objects inside the field.

Add to this ANSI SQL custom types and the necessity to deal with nested fields in the future, and the problem renders really complex. I doubt we will have the proper solution at the moment because we do no know how to deal with nested fields yet. So maybe `ANY` is a good starting point for beta. We may change it in future if needed.

Reply:
Added a user defined type for `OBJECT` as an alias to `ANY`, it's now possible to use `OBJECT` in queries, `ANY` is not allowed in queries anymore, but it's still used internally.

Reply:
Thanks. Will Calcite exceptions `SqlCallBinding#newValidationSignatureError` show `OBJECT` now instead of `ANY`?

Reply:
Added a custom Hazelcast OBJECT type in 7fea4589d426116b8e85e1bb3723dc596568fffa, all references to `ANY` type now should be displayed as `OBJECT`.